src/sources/splunk_hec/acknowledgements.rs

```diff
@@ -0,0 +1,482 @@
+            channels,
+            shutdown,
+            total_pending_acks: AtomicU64::new(0),
+        }
+    }
+
+    /// Creates a channel with the specified id if it does not exist.
+    async fn create_or_get_channel(&self, id: String) -> Result<Arc<Channel>, Rejection> {
+        let mut channels = self.channels.lock().await;
+        if let Some(channel) = channels.get(&id) {
+            return Ok(Arc::clone(channel));
+        }
+
+        if channels.len() < self.max_number_of_ack_channels as usize {
+            // Create the channel if it does not exist
+            let channel = Arc::new(Channel::new(
+                self.max_pending_acks_per_channel,
+                self.shutdown.clone(),
+            ));
+            channels.insert(id, Arc::clone(&channel));
+            Ok(channel)
+        } else {
+            Err(Rejection::from(ApiError::ServiceUnavailable))
+        }
+    }
+
+    /// Gets the next available ack id from a specified channel, creating the channel if it does not exist
+    pub async fn get_ack_id_from_channel(
+        &self,
+        channel_id: String,
+        batch_rx: BatchStatusReceiver,
+    ) -> Result<u64, Rejection> {
+        let channel = self.create_or_get_channel(channel_id).await?;
+        let total_pending_acks = self.total_pending_acks.fetch_add(1, Ordering::Relaxed) + 1;
+        if total_pending_acks > self.max_pending_acks
+            && !self.drop_oldest_pending_ack_from_channels().await
+        {
+            self.total_pending_acks.fetch_sub(1, Ordering::Relaxed);
+            return Err(Rejection::from(ApiError::ServiceUnavailable));
+        }
+
+        let ack_id = channel.get_ack_id(batch_rx);
+        Ok(ack_id)
+    }
+
+    /// Gets the requested ack id statuses from a specified channel, creating the channel if it does not exist
+    pub async fn get_acks_status_from_channel(
+        &self,
+        channel_id: String,
+        ack_ids: &[u64],
+    ) -> Result<HashMap<u64, bool>, Rejection> {
+        let channel = self.create_or_get_channel(channel_id).await?;
+        let acks_status = channel.get_acks_status(ack_ids);
+        let dropped_ack_count = acks_status.values().filter(|status| **status).count();
+        self.total_pending_acks
+            .fetch_sub(dropped_ack_count as u64, Ordering::Relaxed);
+        Ok(acks_status)
+    }
+
+    /// Drops the oldest ack id (if one exists) across all channels
+    async fn drop_oldest_pending_ack_from_channels(&self) -> bool {
+        let channels = self.channels.lock().await;
+        let mut ordered_channels = channels.values().collect::<Vec<_>>();
+        ordered_channels.sort_by_key(|a| a.get_last_used());
+        for channel in ordered_channels {
+            if channel.drop_oldest_pending_ack() {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+pub struct Channel {
+    last_used_timestamp: Mutex<Instant>,
```

Comment:
I believe this would profitably be a [`RwLock`](https://doc.rust-lang.org/std/sync/struct.RwLock.html) instead of a mutex, since you do have a prominent read-only path for this variable.

Reply:
Ah yes, I see that this would be helpful for `get_last_used()`. Will convert.